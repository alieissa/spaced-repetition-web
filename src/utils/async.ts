/** @format */
/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Untriggered = {
  type: 'Untriggered'
}
export function Untriggered(): Untriggered {
  return {
    type: 'Untriggered',
  }
}

const t: Optional<Answer, 'easiness' | 'quality' | 'interval'> = {
  id: '',
  content: 'fsdfds',
}
export type Loading<V> = {
  type: 'Loading'
  value: V
}
export function Loading<V>(v: V): Loading<V> {
  return {
    type: 'Loading',
    value: v,
  }
}

export type Failure<E> = {
  type: 'Failure'
  value: E
}
export function Failure<E>(e: E): Failure<E> {
  return {
    type: 'Failure',
    value: e,
  }
}

export type Success<V> = {
  type: 'Success'
  value: V
}
export function Success<V>(v: V): Success<V> {
  return {
    type: 'Success',
    value: v,
  }
}

export type Async<L = unknown, F = unknown, S = unknown> =
  | Untriggered
  | Loading<L>
  | Failure<F>
  | Success<S>

export type AsyncStatus<R, E, A> = Async<R, E, A>['type']

export type AsyncMap<R, E, A, U> = {
  [K in AsyncStatus<R, E, A>]: U extends { type: K } ? U : never
}

export type AsyncStatusMap<R, E, A> = AsyncMap<R, E, A, Async<R, E, A>>
export type AsyncPattern<T, R, E, A> = {
  [K in keyof AsyncStatusMap<R, E, A>]: (a: AsyncStatusMap<R, E, A>[K]) => T
}

export function match<R, E, A>(
  e: Async<R, E, A>,
): <T>(pattern: AsyncPattern<T, R, E, A>) => T {
  return <T>(pattern: AsyncPattern<T, R, E, A>) => pattern[e.type](e as any)
}
