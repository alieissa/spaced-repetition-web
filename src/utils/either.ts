/** @format */

/**
 * Credit goes to https://medium.com/@fillopeter/pattern-matching-with-typescript-done-right-94049ddd671c
 * For the pattern matching code
 */
export type Right<V> = Readonly<{
  type: 'Right'
  value: V
}>
export function Right<V>(v: V): Right<V> {
  return {
    type: 'Right',
    value: v,
  }
}

export type Left<E = Error> = Readonly<{
  type: 'Left'
  value: E
}>
export function Left<E>(e: E): Left<E> {
  return {
    type: 'Left',
    value: e,
  }
}

export type Either<L, R> = Left<L> | Right<R>

export type EitherType<L, R> = Either<L, R>['type']
export type EitherMap<L, R, U> = {
  [K in EitherType<L, R>]: U extends { type: K } ? U : never
}
export type EitherTypeMap<L, R> = EitherMap<L, R, Either<L, R>>
export type EitherPattern<T, L, R> = {
  [K in keyof EitherTypeMap<L, R>]: (either: EitherTypeMap<L, R>[K]) => T
}

export function match<L, R>(
  e: Either<L, R>,
): <T>(pattern: EitherPattern<T, L, R>) => T {
  return <T>(pattern: EitherPattern<T, L, R>) => pattern[e.type](e as any)
}
