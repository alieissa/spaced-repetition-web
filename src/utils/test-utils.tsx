/** @format */

import { combineReducers, configureStore } from '@reduxjs/toolkit'
import { render } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import { RouterProvider, createMemoryRouter } from 'react-router-dom'
import decksReducer from 'src/modules/decks/decks.reducer'

// As a basic setup, import your same slice reducers
import { loginReducer } from 'src/modules/login'
import { signupReducer } from 'src/modules/signup'
import { verificationReducer } from 'src/modules/verification'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
// interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
//   preloadedState?: PreloadedState<RootState>
//   store?: AppStore
// }

const reducers = combineReducers({
  login: loginReducer,
  signup: signupReducer,
  verification: verificationReducer,
  decks: decksReducer,
})

const withRedux = (ui: React.ReactElement) => {
  const store = configureStore({ reducer: reducers })
  return <Provider store={store}>{ui}</Provider>
}
const withRouter = (ui: React.ReactElement) => {
  const router = createMemoryRouter([{ element: ui, path: '/' }], {
    initialEntries: ['/'],
    initialIndex: 1,
  })

  return <RouterProvider router={router} />
}
export function renderWithProviders(ui: React.ReactElement) {
  // Return an object with the store and all of RTL's query functions
  return render(withRedux(withRouter(ui)))
}

export const flushPromises = () =>
  new Promise((resolve) => setTimeout(resolve, 0))
