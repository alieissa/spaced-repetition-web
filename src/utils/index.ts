/** @format */

import * as async from './async'
import * as either from './either'

export { async }
export { either }
