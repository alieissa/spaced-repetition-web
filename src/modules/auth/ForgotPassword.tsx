import { useFormik } from 'formik'
import { Form } from 'semantic-ui-react'
import { SPButton, SPInput, SPText } from 'src/components'
import { styles } from 'src/styles'
import * as Yup from 'yup'

const ForgotPasswordValidatonSchema = Yup.object().shape({
  email: Yup.string().email('Invalid email').required('Required'),
})
export default function ForgotPassword() {
  const form = useFormik({
    initialValues: {
      email: '',
    },
    validationSchema: ForgotPasswordValidatonSchema,
    onSubmit: (values) => console.log(values),
  })

  const handleOnEmailChange = () => {
    form.setFieldValue('email', form.values.email)
  }
  const emailError = form.touched.email && !!form.errors.email

  return (
    <Form data-testid="forgot-password-form" onSubmit={form.submitForm}>
      <SPInput
        data-testid="forgot-password-form-email-input"
        type="email"
        value={form.values.email}
        onChange={handleOnEmailChange}
      />
      {emailError && (
        <SPText
          data-testid="forgot-password-form-email-error"
          style={styles['mt-0.5r']}
        >
          {form.errors.email}
        </SPText>
      )}
      <SPButton data-testid="forgot-password-form-btn" type="submit">
        Send
      </SPButton>
    </Form>
  )
}
