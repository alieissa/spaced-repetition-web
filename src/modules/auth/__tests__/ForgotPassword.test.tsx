/** @format */

import { faker } from '@faker-js/faker'
import '@testing-library/jest-dom'
import { act, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { renderWithProviders } from 'src/utils/test-utils'
import ForgotPassword from '../ForgotPassword'

const forgotPasswordUrl = `${process.env.REACT_APP_API_ENDPOINT}/users/forgot-password`

const email = faker.internet.email({ firstName: 'Johnny', lastName: 'Bravo' })
export const handlers = [
  rest.post(forgotPasswordUrl, (__, res, ctx) => {
    return res(ctx.json({ email }))
  }),
]

const setupForm = async (fields: 'email'[]) => {
  const user = userEvent.setup()

  if (fields.includes('email')) {
    const emailInput = screen.getByTestId('forgot-password-form-email-input')
    await act(() => user.type(emailInput, email))
  }
}

const submitForm = async () => {
  const forgotPasswordBtn = screen.getByTestId('forgot-password-form-btn')
  await act(() => forgotPasswordBtn.click())
}

const expectElementToBeInDocument = async (testId: string) => {
  const element = await screen.findByTestId(testId)
  await waitFor(() => expect(element).toBeInTheDocument())
}
const server = setupServer(...handlers)
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

describe('ForgotPassword', () => {
  describe('view', () => {
    it('should render correctly', () => {
      const { asFragment } = renderWithProviders(<ForgotPassword />)
      expect(asFragment()).toMatchSnapshot()
    })
    it('should display initial view', async () => {
      renderWithProviders(<ForgotPassword />)
      await waitFor(() => expectElementToBeInDocument('forgot-password-form'))
    })

    // Loading view not displayed so no test necessary
    // it('should display loading view', async ...)

    it('should display error view', async () => {
      // Assemble
      server.use(
        rest.post(forgotPasswordUrl, (__, res, ctx) => {
          return res(ctx.status(422))
        }),
      )
      renderWithProviders(<ForgotPassword />)
      await setupForm(['email'])

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('forgot-password-form-error'),
      )
    })

    it('should display successful submission message', async () => {
      // Assemble
      renderWithProviders(<ForgotPassword />)
      await setupForm(['email'])

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('forgot-password-form-success'),
      )
    })
  })

  describe('interaction', () => {
    it('should display error when email is invalid', async () => {
      // Assemble
      renderWithProviders(<ForgotPassword />)
      await setupForm([])

      // Act
      await submitForm()

      // Assert
      await waitFor(() =>
        expectElementToBeInDocument('forgot-password-form-email-error'),
      )
    })
  })
})
