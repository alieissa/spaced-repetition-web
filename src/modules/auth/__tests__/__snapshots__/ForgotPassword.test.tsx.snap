// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForgotPassword view should render correctly 1`] = `
<DocumentFragment>
  <form
    class="ui form"
    data-testid="forgot-password-form"
  >
    <div
      class="ui input"
      style="opacity: 1;"
    >
      <input
        data-testid="forgot-password-form-email-input"
        style="border-radius: unset;"
        type="email"
        value=""
      />
    </div>
    <button
      class="ui button"
      data-testid="forgot-password-btn"
      style="border-radius: 0;"
      type="submit"
    >
      Send
    </button>
  </form>
</DocumentFragment>
`;
